src/cnav_msg.c

```diff
@@ -0,0 +1,449 @@
+ * preamble sequence is encountered in the message body. For this case, the
+ * function performs for a scan for a preamble with a different offset:
+ * - When found, the preamble octet is moved into the head of the buffer.
+ * - When not found, only 7 bits are left in the buffer.
+ *
+ * \param[in,out] part Decoded component.
+ *
+ * \return None
+ *
+ * \private
+ */
+void _cnav_rescan_preamble(cnav_v27_part_t *part)
+{
+  part->preamble_seen = false;
+
+  if (part->n_decoded > GPS_CNAV_PREAMBLE_LENGTH + 1) {
+    for (size_t i = 1, j = part->n_decoded - GPS_CNAV_PREAMBLE_LENGTH;
+         i < j; ++i) {
+      u32 c = getbitu(part->decoded, i, GPS_CNAV_PREAMBLE_LENGTH);
+      if (GPS_CNAV_PREAMBLE1 == c || GPS_CNAV_PREAMBLE2 == c) {
+        part->preamble_seen = true;
+        part->invert = (GPS_CNAV_PREAMBLE2 == c);
+        /* We shift the accumulated bits to the beginning of the buffer */
+        _cnav_bitlshift(part->decoded,
+                        sizeof(part->decoded),
+                        i);
+        part->n_decoded -= i;
+        break;
+      }
+    }
+  }
+  if (!part->preamble_seen && part->n_decoded >= GPS_CNAV_PREAMBLE_LENGTH) {
+    _cnav_bitlshift(part->decoded,
+                    sizeof(part->decoded),
+                    part->n_decoded - GPS_CNAV_PREAMBLE_LENGTH + 1);
+    part->n_decoded = GPS_CNAV_PREAMBLE_LENGTH - 1;
+  }
+}
+
+/**
+ * Feed a symbol into Viterbi decoder instance.
+ *
+ * The method uses block Viterbi decoder. It first accumulates initial number of
+ * symbols, and after that runs decoding every time the buffer is full. Only
+ * some of the decoded symbols are used.
+ *
+ * \param[in,out] part Decoder object
+ * \param[in]     s    Symbol (0x00 - Hard 0, 0xFF - Hard 1)
+ *
+ * \return None
+ *
+ * \private
+ */
+void _cnav_add_symbol(cnav_v27_part_t *part, u8 ch)
+{
+  part->symbols[part->n_symbols++] = ch;
+
+  if (part->init) {
+    /* Initial step - load more symbols without decoding. */
+    if (part->n_symbols < (GPS_L2C_V27_INIT_BITS + GPS_L2C_V27_DECODE_BITS) * 2) {
+      return;
+    }
+    part->init = false;
+  }
+  else if (part->n_symbols < GPS_L2C_V27_DECODE_BITS * 2) {
+    /* Wait until decoding block is accumulated */
+    return;
+  }
+
+  /* Feed accumulated symbols into the buffer, reset the number of accumulated
+   * symbols. */
+  v27_update(&part->dec, part->symbols, part->n_symbols / 2);
+  part->n_symbols = 0;
+
+  /* Decode N+M bits, where:
+   * - N - Number of bits to put into decoded buffer
+   * - M - Number of bits in the tail to ignore.
+   */
+  unsigned char tmp_bits[ (GPS_L2C_V27_DECODE_BITS + GPS_L2C_V27_DELAY_BITS +
+                           CHAR_BIT - 1) / CHAR_BIT];
+
+  v27_chainback_likely(&part->dec, tmp_bits,
+                       GPS_L2C_V27_DECODE_BITS + GPS_L2C_V27_DELAY_BITS);
+
+  /* Read decoded bits and add them to the decoded buffer */
+  u32 to_add = getbitu(tmp_bits, 0, GPS_L2C_V27_DECODE_BITS);
```

Comment:
This breaks if GPS_L2C_V27_DECODE_BITS > 32. Can we add a comment or an assert or something?